packages/next/src/next-devtools/dev-overlay/segment-explorer.ts

```diff
@@ -0,0 +1,124 @@
+import { useSyncExternalStore } from 'react'
+
+/**
+ * Trie data structure for storing and searching paths
+ *
+ * This can be used to store app router paths and search for them efficiently.
+ * e.g.
+ *
+ * [trie root]
+ *   ├── layout.js
+ *   ├── page.js
+ *   ├── blog
+ *       ├── layout.js
+ *       ├── page.js
+ *       ├── [slug]
+ *          ├── layout.js
+ *          ├── page.js
+ **/
+
+type TrieNode<Value = string> = {
+  value: Value | undefined
+  children: {
+    [key: string]: TrieNode<Value> | undefined
+  }
+}
+
+type Trie<Value = string> = {
+  insert: (value: Value) => void
+  remove: (value: Value) => void
+  getRoot: () => TrieNode<Value>
+}
+
+const listeners = new Set<() => void>()
+const createSegmentTreeStore = (): {
+  subscribe: (callback: () => void) => () => void
+  getSnapshot: () => SegmentTrieNode
+  getServerSnapshot: () => SegmentTrieNode
+} => {
+  // return a store that can be used by useSyncExternalStore
+  return {
+    subscribe: (callback) => {
+      listeners.add(callback)
+      return () => listeners.delete(callback)
+    },
+    getSnapshot: () => {
+      return trie.getRoot()
+    },
+    getServerSnapshot: () => {
+      return trie.getRoot()
+    },
+  }
+}
+
+// TODO: Move the Segment Tree into React State
+const { subscribe, getSnapshot, getServerSnapshot } = createSegmentTreeStore()
+
+function createTrie<Value = string>({
+  getKey = (k) => k as unknown as string,
+}: {
+  getKey: (k: Value) => string
+}): Trie<Value> {
+  const root: TrieNode<Value> = {
+    value: undefined,
+    children: {},
+  }
+
+  function markUpdated() {
+    for (const listener of listeners) {
+      listener()
+    }
+  }
+
+  function insert(value: Value) {
+    let currentNode = root
+    const key = getKey(value)
+    const segments = key.split('/')
```

Comment:
A general-purpose Trie would require a `getCharacters` function instead of `getKey`. The Segment Explorer (or any file system) would then implement `getCharacters` as `value => value.pagePath.split("/")`.

```suggestion
    const segments = getCharacters(value)
```

Reply:
This is also addressed in #80443